plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/ArtifactRetriever.java

```diff
@@ -0,0 +1,217 @@
+                  String result = xpath.evaluate("/metadata/versioning/latest", document);
+                  return new DefaultArtifactVersion(result);
+                }
+              });
+
+  private final LoadingCache<String, NavigableSet<ArtifactVersion>> availableVersions =
+      CacheBuilder.newBuilder()
+          .refreshAfterWrite(4, TimeUnit.HOURS)
+          .build(
+              new CacheLoader<String, NavigableSet<ArtifactVersion>>() {
+
+                @Override
+                public NavigableSet<ArtifactVersion> load(String coordinates) throws Exception {
+                  Document document = getMetadataDocument(coordinates);
+                  XPath xpath = XPathFactory.newInstance().newXPath();
+                  NodeList versionNodes = (NodeList) xpath.evaluate(
+                      "/metadata/versioning/versions/version",
+                      document,
+                      XPathConstants.NODESET);
+                  Builder<ArtifactVersion> versions = ImmutableSortedSet.naturalOrder();
+                  for (int i = 0; i < versionNodes.getLength(); i++) {
+                    String versionString = versionNodes.item(i).getTextContent();
+                    versions.add(new DefaultArtifactVersion(versionString));
+                  }
+                  return versions.build();
+                }
+              });
+
+  /**
+   * @param repositoryUrl the base URL of the maven mirror such as 
+   *     "https://repo1.maven.org/maven2/"
+   * @throws URISyntaxException if the argument is not a valid URL
+   */
+  public ArtifactRetriever(String repositoryUrl) throws URISyntaxException {
+    Preconditions.checkNotNull(repositoryUrl);
+    // check for URL syntax
+    new URI(repositoryUrl);
+    if (!repositoryUrl.endsWith("/")) {
+      repositoryUrl = repositoryUrl + "/";
+    }
+    
+    this.repositoryUrl = repositoryUrl;
+  }
+
+  /**
+   * Retrieve from https://repo1.maven.org/maven2/
+   */
+  public ArtifactRetriever() {
+    this.repositoryUrl = "https://repo1.maven.org/maven2/";
+  }
+
+  /**
+   * Returns the latest published artifact version, or null if there is no such version.
+   */
+  public ArtifactVersion getLatestArtifactVersion(String groupId, String artifactId) {
+    return getLatestIncrementalVersion(idToKey(groupId, artifactId), null);
+  }
+
+  /**
+   * Returns the latest published artifact version in the version range, or null if there is no such
+   * version.
+   */
+  public ArtifactVersion getLatestArtifactVersion(
+      String groupId, String artifactId, VersionRange range) {
+    return getLatestIncrementalVersion(idToKey(groupId, artifactId), range);
+  }
+
+  /**
+   * Returns the latest version of the specified artifact in the version range, or null if there is
+   * no such version.
+   * 
+   * @param coordinates Maven coordinates in the form groupId:artifactId
+   */
+  private ArtifactVersion getLatestIncrementalVersion(String coordinates, VersionRange range) {
```

Comment:
Maven version numbers are in the form of `Maj.Min.Inc`, so this is effectively the latest version within the provided version range.